File: python/scanner/quality.py
from .constants import GRADE_BOUNDARIES


def compute(
    total_loc: int,
    total_comment_lines: int,
    total_functions: int,
    total_function_lines: int,
    total_files: int,
    smell_warnings: int,
    smell_errors: int,
    secret_count: int,
    dangerous_call_count: int,
) -> dict:
    code_smell_score = _code_smell_score(total_functions, smell_warnings, smell_errors)
    security_score   = _security_score(secret_count, dangerous_call_count)
    comment_score    = _comment_score(total_loc, total_comment_lines)
    maint_score      = _maintainability_score(total_functions, total_function_lines, total_files, total_loc)

    composite = round(
        code_smell_score * 0.30 +
        security_score   * 0.25 +
        comment_score    * 0.20 +
        maint_score      * 0.25
    )

    return {
        'score': composite,
        'grade': _grade(composite),
        'components': {
            'code_smell_score':      round(code_smell_score),
            'security_score':        round(security_score),
            'comment_score':         round(comment_score),
            'maintainability_score': round(maint_score),
        },
    }


def _code_smell_score(total_functions: int, warnings: int, errors: int) -> float:
    density = (warnings * 1 + errors * 2) / max(total_functions, 1)
    return max(0.0, 100.0 - density * 50.0)


def _security_score(secret_count: int, dangerous_call_count: int) -> float:
    return max(0.0, 100.0 - secret_count * 20.0 - dangerous_call_count * 10.0)


def _comment_score(total_loc: int, total_comment_lines: int) -> float:
    ratio = total_comment_lines / max(total_loc, 1)
    if ratio < 0.05:
        return ratio / 0.05 * 40.0
    if ratio <= 0.30:
        return 40.0 + (ratio - 0.05) / 0.25 * 60.0
    return max(60.0, 100.0 - (ratio - 0.30) * 200.0)


def _maintainability_score(
    total_functions: int,
    total_function_lines: int,
    total_files: int,
    total_loc: int,
) -> float:
    avg_fn_len = total_function_lines / max(total_functions, 1)
    if avg_fn_len <= 20:
        length_score = 100.0
    elif avg_fn_len <= 50:
        length_score = max(0.0, 100.0 - (avg_fn_len - 20) / 30.0 * 50.0)
    else:
        length_score = max(0.0, 50.0 - (avg_fn_len - 50) / 50.0 * 50.0)

    avg_file_loc = total_loc / max(total_files, 1)
    if avg_file_loc <= 200:
        file_score = 100.0
    elif avg_file_loc <= 500:
        file_score = max(0.0, 100.0 - (avg_file_loc - 200) / 300.0 * 40.0)
    else:
        file_score = max(0.0, 60.0 - (avg_file_loc - 500) / 500.0 * 60.0)

    return length_score * 0.6 + file_score * 0.4


def _grade(score: int) -> str:
    for threshold, letter in GRADE_BOUNDARIES:
        if score >= threshold:
            return letter
    return 'F'

File: javascript/src/quality.js
import { GRADE_BOUNDARIES } from './constants.js';

export function compute({
  totalLoc, totalCommentLines, totalFunctions, totalFunctionLines,
  totalFiles, smellWarnings, smellErrors, secretCount, dangerousCallCount,
}) {
  const codeSmellScore    = computeCodeSmellScore(totalFunctions, smellWarnings, smellErrors);
  const securityScore     = computeSecurityScore(secretCount, dangerousCallCount);
  const commentScore      = computeCommentScore(totalLoc, totalCommentLines);
  const maintScore        = computeMaintainabilityScore(totalFunctions, totalFunctionLines, totalFiles, totalLoc);

  const composite = Math.round(
    codeSmellScore * 0.30 +
    securityScore  * 0.25 +
    commentScore   * 0.20 +
    maintScore     * 0.25
  );

  return {
    score: composite,
    grade: computeGrade(composite),
    components: {
      code_smell_score:      Math.round(codeSmellScore),
      security_score:        Math.round(securityScore),
      comment_score:         Math.round(commentScore),
      maintainability_score: Math.round(maintScore),
    },
  };
}

function computeCodeSmellScore(totalFunctions, warnings, errors) {
  const density = (warnings * 1 + errors * 2) / Math.max(totalFunctions, 1);
  return Math.max(0, 100 - density * 50);
}

function computeSecurityScore(secretCount, dangerousCallCount) {
  return Math.max(0, 100 - secretCount * 20 - dangerousCallCount * 10);
}

function computeCommentScore(totalLoc, totalCommentLines) {
  const ratio = totalCommentLines / Math.max(totalLoc, 1);
  if (ratio < 0.05) return ratio / 0.05 * 40;
  if (ratio <= 0.30) return 40 + (ratio - 0.05) / 0.25 * 60;
  return Math.max(60, 100 - (ratio - 0.30) * 200);
}

function computeMaintainabilityScore(totalFunctions, totalFunctionLines, totalFiles, totalLoc) {
  const avgFnLen = totalFunctionLines / Math.max(totalFunctions, 1);
  let lengthScore;
  if (avgFnLen <= 20) lengthScore = 100;
  else if (avgFnLen <= 50) lengthScore = Math.max(0, 100 - (avgFnLen - 20) / 30 * 50);
  else lengthScore = Math.max(0, 50 - (avgFnLen - 50) / 50 * 50);

  const avgFileLoc = totalLoc / Math.max(totalFiles, 1);
  let fileScore;
  if (avgFileLoc <= 200) fileScore = 100;
  else if (avgFileLoc <= 500) fileScore = Math.max(0, 100 - (avgFileLoc - 200) / 300 * 40);
  else fileScore = Math.max(0, 60 - (avgFileLoc - 500) / 500 * 60);

  return lengthScore * 0.6 + fileScore * 0.4;
}

export function computeGrade(score) {
  for (const [threshold, letter] of GRADE_BOUNDARIES) {
    if (score >= threshold) return letter;
  }
  return 'F';
}
